lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -205,25 +222,44 @@ class VerilogEmitterBase {
 
 } // end anonymous namespace
 
-void VerilogEmitterBase::emitTypePaddedToWidth(Type type, size_t padToWidth,
-                                               Operation *op) {
-  int bitWidth = getBitWidthOrSentinel(type);
-  assert(bitWidth != 0 && "Shouldn't emit zero bit declarations");
+/// Emit a type's packed dimensions, returning the number of characters emitted.
+static size_t emitTypeDims(Type type, llvm::raw_ostream &os, Operation *op) {
+  if (auto inout = type.dyn_cast<rtl::InOutType>()) {
+    return emitTypeDims(inout.getElementType(), os, op);
+  } else if (auto uarray = type.dyn_cast<rtl::UnpackedArrayType>()) {
+    return emitTypeDims(uarray.getElementType(), os, op);
+  }
+
+  size_t emittedWidth = 0;
+  int width;
+  if (auto arrayType = type.dyn_cast<rtl::ArrayType>()) {
+    emittedWidth += emitTypeDims(arrayType.getElementType(), os, op);
+    width = arrayType.getSize();
+  } else {
+    width = getBitWidthOrSentinel(type);
+    assert(width != 0 && "Shouldn't emit zero bit declarations");
+  }
 
-  if (bitWidth == -1) {
-    emitError(op, "value has an unsupported verilog type ") << type;
+  if (width == -1) {
+    op->emitError("value has an unsupported verilog type ") << type;
     os << "<<invalid type>>";
-    return;
+    return 16;
   }
 
-  size_t emittedWidth = 0;
-  if (bitWidth != 1) {
-    os << '[' << (bitWidth - 1) << ":0]";
-    emittedWidth = getPrintedIntWidth(bitWidth - 1) + 4;
+  if (width != 1) {
+    os << '[' << (width - 1) << ":0]";
+    emittedWidth += getPrintedIntWidth(width - 1) + 4;
   }
+  return emittedWidth;
+}
 
+void VerilogEmitterBase::emitTypeDimsPaddedToWidth(Type type, size_t padToWidth,
+                                                   Operation *op) {
+  size_t emittedWidth = emitTypeDims(type, os, op);
   if (emittedWidth < padToWidth)
     os.indent(padToWidth - emittedWidth);
+  if (padToWidth > 0 || emittedWidth > 0) // Space before name.
+    os << ' ';
```

Comment:
What is this check doing?  Why not handle it with the preceding os.indent?  I think this implies an off by one error somewhere.

Reply:
It's to handle the case where padding isn't necessary and the last call didn't output any dims. Now that I broke up this function, the callsites which pass in a constant 0 for padWidth can just call `emitTypeDims` and we can avoid this check.